crates/oxc_linter/src/rules/oxc/no_async_endpoint_handlers.rs

```diff
@@ -0,0 +1,376 @@
+                // Unresolved reference? Nothing we can do.
+                let Some(symbol_id) = handler
+                    .reference_id()
+                    .and_then(|id| ctx.symbols().get_reference(id).symbol_id())
+                else {
+                    return;
+                };
+
+                // Cannot check imported handlers without cross-file analysis.
+                let flags = ctx.symbols().get_flags(symbol_id);
+                if flags.is_import() {
+                    return;
+                }
+
+                let decl_id = ctx.symbols().get_declaration(symbol_id);
+                let decl_node = ctx.nodes().get_node(decl_id);
+                let registered_at = registered_at.or(Some(handler.span));
+                match decl_node.kind() {
+                    AstKind::Function(f) => self.check_function(ctx, registered_at, id_name, f),
+                    AstKind::VariableDeclarator(decl) => {
+                        if let Some(init) = &decl.init {
+                            self.check_endpoint_expr(ctx, id_name, registered_at, init);
+                        }
+                    }
+                    _ => {}
+                }
+            }
+            func if utils::is_endpoint_handler(func) => {
+                match func {
+                    // `app.get('/', (async?) function (req, res) {}`
+                    Expression::FunctionExpression(f) => {
+                        self.check_function(ctx, registered_at, id_name, f);
+                    }
+                    Expression::ArrowFunctionExpression(f) => {
+                        self.check_arrow(ctx, registered_at, id_name, f);
+                    }
+                    _ => unreachable!(),
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn check_function<'a>(
+        &self,
+        ctx: &LintContext<'a>,
+        registered_at: Option<Span>,
+        id_name: Option<&str>,
+        f: &Function<'a>,
+    ) {
+        if !f.r#async {
+            return;
+        }
+
+        let name = f.name().map(|n| n.as_str()).or(id_name);
+        if name.is_some_and(|name| self.is_allowed_name(name)) {
+            return;
+        }
+
+        ctx.diagnostic(no_async_handlers(f.span, registered_at, name));
+    }
+
+    fn check_arrow<'a>(
+        &self,
+        ctx: &LintContext<'a>,
+        registered_at: Option<Span>,
+        id_name: Option<&str>,
+        f: &ArrowFunctionExpression<'a>,
+    ) {
+        if !f.r#async {
+            return;
+        }
+        if id_name.is_some_and(|name| self.is_allowed_name(name)) {
+            return;
+        }
+
+        ctx.diagnostic(no_async_handlers(f.span, registered_at, id_name));
+    }
+
+    fn is_allowed_name(&self, name: &str) -> bool {
+        self.allowed_names.binary_search_by(|allowed| allowed.as_str().cmp(name)).is_ok()
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+    use serde_json::json;
+
+    let pass = vec![
```

Comment:
should we allow if there's a `try` `catch` in the stmt?

Reply:
This could introduce a lot of difficulties with checking _what_ is being caught
```ts
async handle(req, res) {
  try {
   await couldThrow()
  } catch {}
  await couldAlsoThrow()
}
// or this example, which is never caught
async handle(req, res) {
  try {
    couldThrow.then(data => res.json(data))
  } catch {}
}
```
Making sure everything is covered by a try-catch sounds painful...

Reply:
hmm i see. this seems reasonable to skip checking